### DISCUSSION TITLE
Is it possible to use the source generators with CSharpCompilation?

### DISCUSSION THREAD

**Discussion by AndrewKeepCoding** [CommunityToolkit/dotnet#626] [Q&A]
I'm creating dynamic ViewModels with ``INotifyPropertyChanged`` and it works, but I wonder if it's possible to use the source generators from the toolkit.

**Comment by behdad088** [comment]
Hi @AndrewKeepCoding  Yes, it is possible to use the source generators with CSharpCompilation. Here's an example of how you can do it:

```c#

using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.Emit;

// Define your source generator class
public class MySourceGenerator : ISourceGenerator
{
    public void Initialize(GeneratorInitializationContext context)
    {
        // Initialization logic
    }

    public void Execute(GeneratorExecutionContext context)
    {
        // Generation logic
    }
}

// Create a compilation with your source generator
var syntaxTree = CSharpSyntaxTree.ParseText("class MyClass {}");
var compilation = CSharpCompilation.Create("MyCompilation",
    syntaxTrees: new[] { syntaxTree },
    options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary),
    references: new[] { MetadataReference.CreateFromFile(typeof(object).Assembly.Location) },
    new[] { new MySourceGenerator() });

// Emit the compilation and inspect the results
using var ms = new MemoryStream();
var result = compilation.Emit(ms);

if (result.Success)
{
    // Compilation succeeded, do something with the output assembly
    var assembly = Assembly.Load(ms.ToArray());
}
else
{
    // Compilation failed, handle the error
    foreach (var diagnostic in result.Diagnostics)
    {
        Console.Error.WriteLine(diagnostic);
    }
}

``` 
In the above code, we define a source generator class MySourceGenerator, create a C# syntax tree with a simple class definition, create a CSharpCompilation with the syntax tree and the source generator, and emit the compilation to a memory stream. If the compilation succeeds, we load the resulting assembly from the memory stream. If the compilation fails, we print the compilation errors to the console.

**Reply by AndrewKeepCoding** [reply]
I'm so sorry for the late reply. In my question, I meant the CommunityToolkit.Mvvm's source generators.

For example, if we pass the text below and make it also generate a partial class with INotifyPropertyChanged implemented:

```
public partial class SomeViewModel : ObservableObject
{
    [ObservableProperty]
    private double someNumber;
}
```
